packages/sdk/src/chains/EVM/genericMessage.ts

```diff
@@ -0,0 +1,178 @@
+   * Calculates fee required for the requested transfer.
+   * Fee can be paid in native currency or ERC20 token if the 'tokenAddress'
+   * is defined.
+   *
+   * @param transfer instance of transfer
+   * @returns fee that needs to paid
+   */
+  public async getFee(transfer: Transfer<TransferType>): Promise<EvmFee> {
+    const domainConfig = this.config.getDomainConfig() as EthereumConfig;
+    const feeRouter = FeeHandlerRouter__factory.connect(domainConfig.feeRouter, this.provider);
+    const feeHandlerAddress = await feeRouter._domainResourceIDToFeeHandlerAddress(
+      transfer.to.id,
+      transfer.resource.resourceId,
+    );
+    const feeHandlerConfig = domainConfig.feeHandlers.find(
+      feeHandler => feeHandler.address == feeHandlerAddress,
+    )!;
+
+    switch (feeHandlerConfig.type) {
+      case FeeHandlerType.BASIC: {
+        return await calculateBasicfee({
+          basicFeeHandlerAddress: feeHandlerAddress,
+          provider: this.provider,
+          fromDomainID: transfer.from.id,
+          toDomainID: transfer.to.id,
+          resourceID: transfer.resource.resourceId,
+          sender: transfer.sender,
+        });
+      }
+      case FeeHandlerType.DYNAMIC: {
+        return await calculateDynamicFee({
+          provider: this.provider,
+          sender: transfer.sender,
+          fromDomainID: Number(transfer.from.id),
+          toDomainID: Number(transfer.to.id),
+          resourceID: transfer.resource.resourceId,
+          feeOracleBaseUrl: getFeeOracleBaseURL(this.environment),
+          feeHandlerAddress: feeHandlerAddress,
+          depositData: createPermissionlessGenericDepositData(
+            (transfer as Transfer<GenericMessage>).details.executeFunctionSignature,
+            (transfer as Transfer<GenericMessage>).details.executeContractAddress,
+            (transfer as Transfer<GenericMessage>).details.maxFee,
+            transfer.sender,
+            (transfer as Transfer<GenericMessage>).details.executionData,
+          ),
+        });
+      }
+      default:
+        throw new Error(`Unsupported fee handler type`);
+    }
+  }
+
+  /**
+   * Builds an unsigned transfer transaction.
+   * Should be executed after the approval transactions.
+   *
+   * @param transfer
+   * @param fee
+   * @returns unsigned transfer transaction
+   */
+  public async buildTransferTransaction(
+    transfer: Transfer<TransferType>,
+    fee: EvmFee,
+  ): Promise<PopulatedTransaction> {
+    const bridge = Bridge__factory.connect(this.config.getDomainConfig().bridge, this.provider);
+    switch (transfer.resource.type) {
+      case ResourceType.PERMISSIONLESS_GENERIC: {
+        return await genericMessageTransfer({
+          depositor: transfer.sender,
+          bridgeInstance: bridge,
+          domainId: transfer.to.id.toString(),
+          resourceId: transfer.resource.resourceId,
+          feeData: fee,
+          executeContractAddress: (transfer as Transfer<GenericMessage>).details
```

Comment:
Can you cast somewhere earlier so you don't have 4 casts to Transfer<GenericMessage>